public/app/features/alerting/unified/components/rules/state-history/LokiStateHistory.tsx

```diff
@@ -0,0 +1,356 @@
+    stateHistory,
+    instancesFilter
+  );
+
+  const frameSubset = useMemo(() => take(dataFrames, MAX_TIMELINE_SERIES), [dataFrames]);
+
+  const onLogRecordLabelClick = useCallback(
+    (label: string) => {
+      const matcherString = combineMatcherStrings(getValues('query'), label);
+      setInstancesFilter(matcherString);
+      setValue('query', matcherString);
+    },
+    [setInstancesFilter, setValue, getValues]
+  );
+
+  const onFilterCleared = useCallback(() => {
+    setInstancesFilter('');
+    setValue('query', '');
+  }, [setInstancesFilter, setValue]);
+
+  const onTimelinePointerMove = useCallback(
+    (seriesIdx: number, pointIdx: number) => {
+      const uniqueTimestamps = sortBy(uniq(frameSubset.flatMap((frame) => frame.fields[0].values.toArray())));
+
+      const timestamp = uniqueTimestamps[pointIdx];
+
+      const refToScroll = logsRef.current.find((x) => parseInt(x.id, 10) === timestamp);
+      refToScroll?.scrollIntoView({ behavior: 'smooth', block: 'start' });
+    },
+    [frameSubset]
+  );
+
+  if (isLoading) {
+    return <div>Loading...</div>;
+  }
+  if (isError) {
+    return (
+      <Alert title="Error fetching the state history" severity="error">
+        {error instanceof Error ? error.message : 'Unable to fetch alert state history'}
+      </Alert>
+    );
+  }
+
+  const hasMoreInstances = frameSubset.length < dataFrames.length;
+  const emptyStateMessage =
+    totalRecordsCount > 0
+      ? `No matches were found for the given filters among the ${totalRecordsCount} instances`
+      : 'No state transitions have occurred in the last 60 minutes';
+
+  return (
+    <div className={styles.fullSize}>
+      <form onSubmit={handleSubmit((data) => setInstancesFilter(data.query))}>
+        <Field
+          label={
+            <Label htmlFor="instancesSearchInput">
+              <Stack gap={0.5}>
+                <span>Filter instances</span>
+                <HoverCard
+                  content={
+                    <>
+                      Use label matcher expression (like <code>{'{foo=bar}'}</code>) or click on an instance label to
+                      filter instances
+                    </>
+                  }
+                >
+                  <Icon name="info-circle" size="sm" />
+                </HoverCard>
+              </Stack>
+            </Label>
+          }
+        >
+          <Input
+            {...register('query')}
+            id="instancesSearchInput"
+            prefix={<Icon name="search" />}
+            suffix={
+              <Button fill="text" icon="times" size="sm" onClick={onFilterCleared}>
+                Clear
+              </Button>
+            }
+            placeholder="Filter instances"
+          />
+        </Field>
+        <input type="submit" hidden />
+      </form>
+      {!isEmpty(commonLabels) && (
+        <Stack direction="row" alignItems="center">
+          <strong>Common labels</strong>
```

Comment:
Might not be bad to have a tooltip explaining what a common label is (the ones that occur in every record).